File: plugins/plugin-telegram/python/elizaos_plugin_telegram/actions/send_reaction.py
"""Send reaction action for Telegram plugin."""
import logging
from dataclasses import dataclass

from elizaos_plugin_telegram.types import (
    SendReactionParams,
    SendReactionResult,
    TelegramReactions,
)

logger = logging.getLogger(__name__)


@dataclass
class SendReactionAction:
    """Action to send a reaction to a Telegram message."""

    name: str = "SEND_TELEGRAM_REACTION"
    description: str = "Send a reaction emoji to a Telegram message"
    similes: tuple[str, ...] = (
        "TELEGRAM_REACT",
        "TELEGRAM_REACTION",
        "REACT_TO_MESSAGE",
        "ADD_REACTION",
        "SEND_EMOJI",
        "TELEGRAM_EMOJI",
    )

    def validate(self, source: str | None) -> bool:
        """Validate that the action can be performed."""
        return source == "telegram"

    async def handler(
        self,
        service: object,
        chat_id: int | str,
        message_id: int,
        reaction: str,
        is_big: bool = False,
    ) -> SendReactionResult:
        """Handle the send reaction action.
        
        Args:
            service: The TelegramService instance.
            chat_id: The chat ID where the message is.
            message_id: The ID of the message to react to.
            reaction: The emoji to use as a reaction.
            is_big: Whether to use a big/animated reaction.
            
        Returns:
            SendReactionResult with success status and details.
        """
        # Import here to avoid circular imports
        from elizaos_plugin_telegram.service import TelegramService
        
        if not isinstance(service, TelegramService):
            return SendReactionResult(
                success=False,
                chat_id=chat_id,
                message_id=message_id,
                reaction=reaction,
                error="Invalid service type",
            )
        
        params = SendReactionParams(
            chat_id=chat_id,
            message_id=message_id,
            reaction=reaction,
            is_big=is_big,
        )
        
        return await service.send_reaction(params)


# Map common reaction names to emojis
REACTION_NAME_MAP: dict[str, str] = {
    "thumbs_up": TelegramReactions.THUMBS_UP,
    "thumbsup": TelegramReactions.THUMBS_UP,
    "like": TelegramReactions.THUMBS_UP,
    "+1": TelegramReactions.THUMBS_UP,
    "thumbs_down": TelegramReactions.THUMBS_DOWN,
    "thumbsdown": TelegramReactions.THUMBS_DOWN,
    "dislike": TelegramReactions.THUMBS_DOWN,
    "-1": TelegramReactions.THUMBS_DOWN,
    "heart": TelegramReactions.HEART,
    "love": TelegramReactions.HEART,
    "fire": TelegramReactions.FIRE,
    "lit": TelegramReactions.FIRE,
    "hot": TelegramReactions.FIRE,
    "celebration": TelegramReactions.CELEBRATION,
    "party": TelegramReactions.CELEBRATION,
    "tada": TelegramReactions.CELEBRATION,
    "crying": TelegramReactions.CRYING,
    "sad": TelegramReactions.CRYING,
    "thinking": TelegramReactions.THINKING,
    "hmm": TelegramReactions.THINKING,
    "exploding_head": TelegramReactions.EXPLODING_HEAD,
    "mindblown": TelegramReactions.EXPLODING_HEAD,
    "screaming": TelegramReactions.SCREAMING,
    "scared": TelegramReactions.SCREAMING,
    "angry": TelegramReactions.ANGRY,
    "skull": TelegramReactions.SKULL,
    "dead": TelegramReactions.SKULL,
    "poop": TelegramReactions.POOP,
    "clown": TelegramReactions.CLOWN,
    "eyes": TelegramReactions.EYES,
    "look": TelegramReactions.EYES,
    "hundred": TelegramReactions.HUNDRED,
    "100": TelegramReactions.HUNDRED,
    "perfect": TelegramReactions.HUNDRED,
    "tears_of_joy": TelegramReactions.TEARS_OF_JOY,
    "lol": TelegramReactions.TEARS_OF_JOY,
    "laugh": TelegramReactions.TEARS_OF_JOY,
    "lightning": TelegramReactions.LIGHTNING,
    "zap": TelegramReactions.LIGHTNING,
    "trophy": TelegramReactions.TROPHY,
    "win": TelegramReactions.TROPHY,
    "winner": TelegramReactions.TROPHY,
    "broken_heart": TelegramReactions.BROKEN_HEART,
    "heartbroken": TelegramReactions.BROKEN_HEART,
    "ghost": TelegramReactions.GHOST,
    "boo": TelegramReactions.GHOST,
    "unicorn": TelegramReactions.UNICORN,
}


def normalize_reaction(reaction: str) -> str:
    """Normalize a reaction string to an emoji.
    
    If the input is already an emoji, return it as-is.
    If it's a name, look it up in the map.
    """
    if not reaction:
        return TelegramReactions.THUMBS_UP
    
    # Already an emoji (check if it starts with a non-ASCII char)
    if ord(reaction[0]) > 127:
        return reaction
    
    # Look up by name (case-insensitive)
    normalized = reaction.lower().replace(" ", "_").replace("-", "_")
    return REACTION_NAME_MAP.get(normalized, TelegramReactions.THUMBS_UP)


# Action instance for export
send_reaction_action = SendReactionAction()

File: plugins/plugin-telegram/rust/src/actions/send_reaction.rs
//! Send reaction action for Telegram plugin.

use serde::{Deserialize, Serialize};

use crate::service::TelegramService;
use crate::types::{reactions, SendReactionParams, SendReactionResult};

/// Action name constant.
pub const SEND_REACTION_ACTION: &str = "SEND_TELEGRAM_REACTION";

/// Parameters for the send reaction action.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SendReactionActionParams {
    /// Chat ID where the message is.
    pub chat_id: i64,
    /// Message ID to react to.
    pub message_id: i64,
    /// Reaction to send (emoji or name).
    pub reaction: String,
    /// Whether to use big/animated reaction.
    #[serde(default)]
    pub is_big: bool,
}

/// Execute the send reaction action.
pub async fn execute_send_reaction(
    service: &TelegramService,
    params: SendReactionActionParams,
) -> SendReactionResult {
    let normalized_reaction = normalize_reaction(&params.reaction);
    
    service
        .send_reaction(SendReactionParams {
            chat_id: params.chat_id,
            message_id: params.message_id,
            reaction: normalized_reaction,
            is_big: params.is_big,
        })
        .await
}

/// Normalize a reaction string to an emoji.
///
/// If the input is already an emoji, returns it as-is.
/// If it's a name, looks it up in the common reactions map.
pub fn normalize_reaction(reaction: &str) -> String {
    if reaction.is_empty() {
        return reactions::THUMBS_UP.to_string();
    }

    // Check if it's already an emoji (non-ASCII first char)
    if reaction.chars().next().map_or(false, |c| c as u32 > 127) {
        return reaction.to_string();
    }

    // Normalize the name
    let normalized = reaction.to_lowercase().replace([' ', '-'], "_");

    // Look up in the map
    match normalized.as_str() {
        "thumbs_up" | "thumbsup" | "like" | "+1" => reactions::THUMBS_UP,
        "thumbs_down" | "thumbsdown" | "dislike" | "-1" => reactions::THUMBS_DOWN,
        "heart" | "love" => reactions::HEART,
        "fire" | "lit" | "hot" => reactions::FIRE,
        "celebration" | "party" | "tada" => reactions::CELEBRATION,
        "crying" | "sad" => reactions::CRYING,
        "thinking" | "hmm" => reactions::THINKING,
        "exploding_head" | "mindblown" => reactions::EXPLODING_HEAD,
        "screaming" | "scared" => reactions::SCREAMING,
        "angry" => reactions::ANGRY,
        "skull" | "dead" => reactions::SKULL,
        "poop" => reactions::POOP,
        "clown" => reactions::CLOWN,
        "eyes" | "look" => reactions::EYES,
        "hundred" | "100" | "perfect" => reactions::HUNDRED,
        "tears_of_joy" | "lol" | "laugh" => reactions::TEARS_OF_JOY,
        "lightning" | "zap" => reactions::LIGHTNING,
        "trophy" | "win" | "winner" => reactions::TROPHY,
        "broken_heart" | "heartbroken" => reactions::BROKEN_HEART,
        "ghost" | "boo" => reactions::GHOST,
        "unicorn" => reactions::UNICORN,
        _ => reactions::THUMBS_UP,
    }
    .to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_normalize_reaction_emoji() {
        assert_eq!(normalize_reaction("👍"), "👍");
        assert_eq!(normalize_reaction("❤"), "❤");
    }

    #[test]
    fn test_normalize_reaction_name() {
        assert_eq!(normalize_reaction("thumbs_up"), reactions::THUMBS_UP);
        assert_eq!(normalize_reaction("thumbsup"), reactions::THUMBS_UP);
        assert_eq!(normalize_reaction("like"), reactions::THUMBS_UP);
        assert_eq!(normalize_reaction("heart"), reactions::HEART);
        assert_eq!(normalize_reaction("love"), reactions::HEART);
        assert_eq!(normalize_reaction("fire"), reactions::FIRE);
        assert_eq!(normalize_reaction("FIRE"), reactions::FIRE);
    }

    #[test]
    fn test_normalize_reaction_unknown() {
        assert_eq!(normalize_reaction("unknown"), reactions::THUMBS_UP);
        assert_eq!(normalize_reaction(""), reactions::THUMBS_UP);
    }
}
